### PR TITLE
ARROW-10881: [C++] Fix EXC_BAD_ACCESS in PutSpaced

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by emkornfield** [comment]
Is it possible to add a unit test?

**Comment by xhochy** [comment]
@github-actions autotune

**Comment by nealrichardson** [comment]
@xhochy autotune isn't working at the moment--INFRA has blocked some of the Actions we use, including on that workflow. See https://issues.apache.org/jira/browse/INFRA-21239

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -110,11 +110,16 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
...
-                                                                 this->memory_pool()));
-    T* data = reinterpret_cast<T*>(buffer->mutable_data());
-    int num_valid_values = ::arrow::util::internal::SpacedCompress<T>(
-        src, num_values, valid_bits, valid_bits_offset, data);
+    const T* data = src;
+    int num_valid_values = num_values;
+    if (valid_bits != NULLPTR) {
+      PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
+                                                                   this->memory_pool()));
+      T* data_ = reinterpret_cast<T*>(buffer->mutable_data());
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:118]: Underscore-terminated variable names are only for attributes. Can you use another name? (e.g. `non_spaced_data`)

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -110,11 +110,16 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
...
-    PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
-                                                                 this->memory_pool()));
-    T* data = reinterpret_cast<T*>(buffer->mutable_data());
-    int num_valid_values = ::arrow::util::internal::SpacedCompress<T>(
-        src, num_values, valid_bits, valid_bits_offset, data);
+    const T* data = src;
+    int num_valid_values = num_values;
+    if (valid_bits != NULLPTR) {
+      PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
+                                                                   this->memory_pool()));
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:117]: It seems `buffer` will be deallocated at the end of this `if` block, but you're using its contents afterwards (when calling `Put`).

File: cpp/src/parquet/encoding.cc [hunk 3]
```diff
@@ -309,11 +314,16 @@ class PlainEncoder<BooleanType> : public EncoderImpl, virtual public BooleanEnco
...
                  int64_t valid_bits_offset) override {
-    PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
-                                                                 this->memory_pool()));
-    T* data = reinterpret_cast<T*>(buffer->mutable_data());
-    int num_valid_values = ::arrow::util::internal::SpacedCompress<T>(
-        src, num_values, valid_bits, valid_bits_offset, data);
+    const T* data = src;
+    int num_valid_values = num_values;
+    if (valid_bits != NULLPTR) {
+      PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:320]: Same concerns below.

File: cpp/src/parquet/encoding.h [hunk 4]
```diff
@@ -290,14 +290,18 @@ class TypedDecoder : virtual public Decoder {
...
+      int values_to_read = num_values - null_count;
+      int values_read = Decode(buffer, values_to_read);
+      if (values_read != values_to_read) {
+        throw ParquetException("Number of values / definition_levels read did not match");
+      }
+
+      return ::arrow::util::internal::SpacedExpand<T>(buffer, num_values, null_count,
+                                                      valid_bits, valid_bits_offset);
+    } else {
+      return num_values;
```
Comment (pitrou) [inline cpp/src/parquet/encoding.h:303]: Why don't you call `Decode` in this branch?

**Review by pitrou — CHANGES_REQUESTED** [review]
It would indeed be nice if there was a test.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
CI tests are green on my fork (except for the usual Homebrew dependency install failures).

**Comment by xhochy** [comment]
@pitrou Thanks!

**Pull request closed without merging**